File: haskell/day23_part1_2016.hs
import System.IO
import Data.List
import Data.Maybe
import Control.Monad
import qualified Data.Map as Map

data Instruction = Cpy String String | Inc String | Dec String | Jnz String String | Tgl String deriving (Show)

type Registers = Map.Map String Int

main = do
    contents <- readFile "input.txt"
    let instructions = map parseInstruction (lines contents)
    let registers = Map.fromList [("a", 7), ("b", 0), ("c", 0), ("d", 0)]
    let finalRegisters = executeInstructions instructions registers 0
    print $ Map.findWithDefault 0 "a" finalRegisters

parseInstruction :: String -> Instruction
parseInstruction line =
    case words line of
        ["cpy", x, y] -> Cpy x y
        ["inc", x] -> Inc x
        ["dec", x] -> Dec x
        ["jnz", x, y] -> Jnz x y
        ["tgl", x] -> Tgl x

executeInstructions :: [Instruction] -> Registers -> Int -> Registers
executeInstructions instructions registers pc
    | pc < 0 || pc >= length instructions = registers
    | otherwise =
        case instructions !! pc of
            Cpy x y -> executeInstructions instructions (cpy x y registers) (pc + 1)
            Inc x -> executeInstructions instructions (inc x registers) (pc + 1)
            Dec x -> executeInstructions instructions (dec x registers) (pc + 1)
            Jnz x y -> executeInstructions instructions registers (pc + jnz x y registers)
            Tgl x -> executeInstructions (tgl x pc instructions registers) registers (pc + 1)

cpy :: String -> String -> Registers -> Registers
cpy x y registers
    | Map.member y registers = Map.insert y (getValue x registers) registers
    | otherwise = registers

inc :: String -> Registers -> Registers
inc x registers = Map.insert x (Map.findWithDefault 0 x registers + 1) registers

dec :: String -> Registers -> Registers
dec x registers = Map.insert x (Map.findWithDefault 0 x registers - 1) registers

jnz :: String -> String -> Registers -> Int
jnz x y registers
    | getValue x registers /= 0 = getValue y registers
    | otherwise = 1

tgl :: String -> Int -> [Instruction] -> Registers -> [Instruction]
tgl x pc instructions registers =
    let target = pc + getValue x registers
    in if target >= 0 && target < length instructions
        then let toggled = toggle (instructions !! target)
             in take target instructions ++ [toggled] ++ drop (target + 1) instructions
        else instructions

toggle :: Instruction -> Instruction
toggle (Inc x) = Dec x
toggle (Dec x) = Inc x
toggle (Tgl x) = Inc x
toggle (Jnz x y) = Cpy x y
toggle (Cpy x y) = Jnz x y

getValue :: String -> Registers -> Int
getValue x registers = fromMaybe (read x) (Map.lookup x registers)

File: training_data/reflection-tuning/day23_part1_2016.py
def execute_program(instructions):
    registers = {'a': 7, 'b': 0, 'c': 0, 'd': 0}
    pc = 0

    def get_value(x):
        return registers[x] if x in registers else int(x)

    def toggle_instruction(index):
        if 0 <= index < len(instructions):
            parts = instructions[index].split()
            if len(parts) == 2:
                parts[0] = 'dec' if parts[0] == 'inc' else 'inc'
            elif len(parts) == 3:
                parts[0] = 'cpy' if parts[0] == 'jnz' else 'jnz'
            instructions[index] = ' '.join(parts)

    while 0 <= pc < len(instructions):
        parts = instructions[pc].split()
        op = parts[0]

        if op == 'cpy':
            if parts[2] in registers:
                registers[parts[2]] = get_value(parts[1])
        elif op == 'inc':
            if parts[1] in registers:
                registers[parts[1]] += 1
        elif op == 'dec':
            if parts[1] in registers:
                registers[parts[1]] -= 1
        elif op == 'jnz':
            if get_value(parts[1]) != 0:
                pc += get_value(parts[2]) - 1
        elif op == 'tgl':
            toggle_instruction(pc + get_value(parts[1]))

        pc += 1

    return registers['a']

with open('input.txt', 'r') as file:
    instructions = file.read().splitlines()

result = execute_program(instructions)
print(result)
